### PR TITLE
Propagate changes of output value through Updateable::update() and Circuit::tick()

### PR DESCRIPTION
**Pull request by dopsi**
See #3

### CODE REVIEW

File: examples/fulladder.rs [hunk 1]
```diff
@@ -110,17 +110,29 @@ fn main() {
         y.replace(triple[1]);
         c.replace(triple[2]);
 
-        for _ in 0..3 {
-            circuit.tick();
+        let mut cycle_count = 0_u32;
```
Comment (hellow554) [inline examples/fulladder.rs]: I (personally) don't like the `_u32` suffix. I wonder if it's really necessary here to give such a type hint or if Rust can deduce it by itself (probably to a `i32`). Can you remove it please?

File: examples/fulladder.rs [hunk 2]
```diff
@@ -110,17 +110,29 @@ fn main() {
...
 
+        // If the number of required ticks is known, it is possible to use
+        // a loop with a fixed number of iterations
+        // ```
+        // for _ in 0..3 {
+        //     circuit.tick();
+        // }
+        // ```
+
+        cycle_count += 1; // To get the correct number of cycles
```
Comment (hellow554) [inline examples/fulladder.rs]: Why not starting with `1` in the first place or using `cycle_count + 1` below in the `println!`?

File: src/circuit.rs [hunk 3]
```diff
@@ -1,17 +1,53 @@
...
+    /// Returns `true` as long as at least one element's output value has changed in the circuit.
+    ///
+    /// ```
+    /// use logical::{Circuit, Signal};
+    /// use logical::models::gates::OrGate;
+    ///
+    /// let mut circuit = Circuit::default();
+    /// /* Configure updaters */
+    ///
+    /// let mut clock_cycles = 0_u32;
```
Comment (hellow554) [inline src/circuit.rs]: Same here with the `_u32`

File: src/circuit.rs [hunk 4]
```diff
@@ -1,17 +1,53 @@
...
+        self.updater.iter_mut().fold(false, |acc, u| acc | u.update())
     }
 
+    /// Add an [`Updateable`](Updateable) to the `Circuit`
+    ///
+    /// ```
+    /// use logical::{Circuit, Signal, Ieee1164};
+    /// use logical::models::gates::OrGate;
+    ///
+    /// let mut sig: logical::Signal<Ieee1164> = Signal::default();
```
Comment (hellow554) [inline src/circuit.rs:41]: `let mut sig: Signal<Ieee1164> = Signal::default();`

**Review by hellow554 — CHANGES_REQUESTED** [review]

**Comment by hellow554** [comment]
https://github.com/hellow554/logical-rs/pull/6/files#diff-a6bffa6b46c2d189ffda6e7545e34e47R16

I'm not happy with this 🤔 You use it to determine, whether the value has been updated or not, correct? What if you instead use 

```
fn update(&mut self) -> bool {
    let new_value = $func(self.a.value());
    let old_value = self.z.replace(new_value);
    old_value != new_value
}
```

Because the values (in this case) are `Ieee1164` types, they are at least cloneable (copyable in this case)  this should work.

File: src/circuit.rs (hunk 4 above)
Comment (dopsi) [reply src/circuit.rs:41]: Indeed

File: examples/fulladder.rs (hunk 2 above)
Comment (dopsi) [reply examples/fulladder.rs]: I'll fix that

**Comment by dopsi** [comment]
You were correct that I used it to get the value before the change.

I now use a PortConnector, is that OK? It does not expose the internal API from inner port.

**Comment by hellow554** [comment]
> I now use a PortConnector, is that OK?

I see why you are doing it... I think I'm fine with accepting this currently, **but** I want a better solution for this.

**Comment by hellow554** [comment]
Closes #3

**Comment by dopsi** [comment]
> I see why you are doing it... I think I'm fine with accepting this currently, **but** I want a better solution for this.

Would changing the signature of the closure from `FnOnce(&mut T)` to `FnOnce(&mut T) -> (T, T)` [here](https://github.com/hellow554/logical-rs/blob/master/src/port/pport.rs#L188)? In the return value, the first element in the tuple is the old value and the second the new value. The `with_value_mut` signature would be changed accordingly to

```rust
pub fn with_value_mut<F: FnOnce(&mut T) -> (T, T)>(&mut self, f: F) -> (T, T);
```

The other way would be to determine if the change happened inside the closure (having it return a `bool` and simply use it as the return value of the `update` function) but I feel this way is less generic and hand over the detection to the crate user instead of handling it internally.

**Comment by hellow554** [comment]
I thought about that at first glance as well, **but** `T` is not restricted in any way, therefore not even clone. This is done intentionally and I don't want to break that guarantee. 
Instead one should/could do things in the closure itself, e.g. cloning the inner value, because the closure knows what type is inside and whether it is clone or not.

**Comment by dopsi** [comment]
I think we can then force the closure to notify the `with_value_mut` function where the value changed or not.

**Pull request merged**